src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,443 @@
+
+    private val iceAgent = Agent(Config.ufragPrefix(), logger).apply {
+        appendHarvesters(this)
+        isControlling = controlling
+        performConsentFreshness = true
+        addStateChangeListener(this@IceTransport::iceStateChanged)
+    }.also {
+        logger.addContext("local_ufrag", it.localUfrag)
+    }
+
+    // TODO: Do we still need the id here now that we have logContext?
+    private val iceStream = iceAgent.createMediaStream("stream-$id").apply {
+        addPairChangeListener(this@IceTransport::iceStreamPairChanged)
+    }
+
+    private val iceComponent = iceAgent.createComponent(
+            iceStream,
+            Transport.UDP,
+            -1,
+            -1,
+            -1,
+            Config.keepAliveStrategy(),
+            Config.useComponentSocket()
+    )
+
+    private val packetStats = PacketStats()
+
+    val icePassword: String
+        get() = iceAgent.localPassword
+
+    /**
+     * Tell this [IceTransport] to start ICE connectivity establishment.
+     */
+    fun startConnectivityEstablishment(transportPacketExtension: IceUdpTransportPacketExtension) {
+        if (!running.get()) {
+            logger.warn("Not starting connectivity establishment, transport is not running")
+            return
+        }
+        if (iceAgent.state.isEstablished) {
+            logger.cdebug { "Connection already established" }
+            return
+        }
+        logger.cdebug { "Starting ICE connectivity establishment" }
+
+        // Set the remote ufrag/password
+        iceStream.remoteUfrag = transportPacketExtension.ufrag
+        iceStream.remotePassword = transportPacketExtension.password
+
+        // If ICE is running already, we try to update the checklists with the
+        // candidates. Note that this is a best effort.
+        val iceAgentStateIsRunning = IceProcessingState.RUNNING == iceAgent.state
+
+        val remoteCandidates = transportPacketExtension.getChildExtensionsOfType(CandidatePacketExtension::class.java)
+        if (iceAgentStateIsRunning && remoteCandidates.isEmpty()) {
+            logger.cdebug { "Ignoring transport extensions with no candidates, " +
+                    "the Agent is already running." }
+            return
+        }
+
+        val remoteCandidateCount = addRemoteCandidates(remoteCandidates, iceAgentStateIsRunning)
+        if (iceAgentStateIsRunning) {
+            when (remoteCandidateCount) {
+                0 -> {
+                    // XXX Effectively, the check above but realizing that all
+                    // candidates were ignored:
+                    // iceAgentStateIsRunning && candidates.isEmpty().
+                }
+                else -> iceComponent.updateRemoteCandidates()
+            }
+        } else if (remoteCandidateCount != 0) {
+            // Once again, because the ICE Agent does not support adding
+            // candidates after the connectivity establishment has been started

```

Comment:
??? Really?  It seems like we do that a lot.

Reply:
This is a copy-paste of some existing logic (including the comment).  But yeah I'm not sure if this still applies.